Memory, speed and synchronization improvements for lexicons

Craigacp opened the pull request:
This is the remainder of the commits which reduce the memory requirements and increase the construction speed of the lexicons. The memory requirements are about a third to a half of the current factorie version, and the runtime has improved greatly as it doesn't cause multiple stop the world GC passes during lexicon construction. There are a couple of other improvements to the construction speed as it creates fewer objects now.

I've added synchronisation to all of the modification operations on a lexicon so it should be safe for multithreading whilst it's being constructed. There is also a method on StackedChainNER which constructs the lexicons a priori, and running that in a single thread before multithreaded computation guarantees that there won't be issues. Similar methods should probably be added to the other NER models (or just a method which constructs all the lexicons on the ResourceLexicon class). Scala's lazy val semantics interacts poorly with multithreading.

There are also some style cleanups to make it fit better with other factorie code, and a bugfix for an off by one error in the findMentions method.

Adam

strubell commented on src/main/scala/cc/factorie/app/nlp/lexicon/AhoCorasick.scala:
```diff
@@ -1,15 +1,5 @@
-   See the License for the specific language governing permissions and
-   limitations under the License.
+/*
+ * Copyright C 2014, 2015, Oracle and/or its affiliates. All rights reserved. 
```
Can you change this and all copyright notices back to the original factorie ones?

Craigacp replied:
No. Our lawyers say this has to stay. We've file the contributors agreement which gives UMass & Andrew a license to our contributions but they are still our copyright.

lvilnis replied:
But this file was already in factorie...

Craigacp replied:
It was a file I'd contributed in an earlier pull request. When we went through the process it was the whole AhoCorasick implementation that went through our lawyers.

strubell commented on src/main/scala/cc/factorie/app/nlp/lexicon/AhoCorasick.scala:
```diff
@@ -26,196 +16,185 @@ import scala.collection.JavaConversions._
+    }
+    var i = 0
+    var curNode = root
+    while (i < input.length) {
```
This loop looks exactly the same as the one in findMentions, above. Can this be pulled out into a helper method or something?

Craigacp replied:
It's not easy to do as the inputs are different and there is mutable state in the loop that needs to be preserved. It might be possible to do it by making a helper method take a function which unwraps the input and a function which tags or collects the mentions, but that would probably be longer and less clear.

strubell commented on src/main/scala/cc/factorie/app/nlp/lexicon/ResourceLexicons.scala:
```diff
@@ -28,11 +28,13 @@ class ResourceLexicons(val sourceFactory: String=>io.Source, val tokenizer:Strin
     this ++= sourceFactory(dir + "/" + name + ".txt")
   }
+
   /** deprecated **/
```
Do we want to @deprecate (i.e. actually deprecate) these?

Craigacp replied:
I was just replicating the comment Kate put above the ChainPhraseLexicon on the PhraseLexicon. I can't think of a reason to use the PhraseLexicon, but the TriePhraseLexicon and ChainPhraseLexicon have different token annotation methods (though the ChainPhraseLexicon is much slower at annotation time).

strubell commented on src/main/scala/cc/factorie/util/JavaHashMap.scala:
```diff
@@ -19,8 +19,10 @@ import scala.collection.JavaConverters._
 // Replace this with AnyRefMap when we switch to Scala 2.11 -luke
 object JavaHashMap {
   def apply[K, V](): mutable.Map[K, V] = new java.util.HashMap[K, V]().asScala
+  def apply[K, V](size : Int): mutable.Map[K, V] = new java.util.HashMap[K, V](size).asScala
```
Nice addition, thanks!

strubell commented:
This looks great, thank you so much! See minimal style/licensing comments inline.

Craigacp commented:
I added the Apache license text back, but the copyright statement is still Oracle. Our lawyers assumed you'd add the license text after accepting the contribution, but that's not really how GitHub works so I put it back in.

I removed the replicated code between tagMentions and lemmatizeAndTagMentions, but I can't do it with tagMentions and findMentions as the type relations got nasty. Maybe someone with better Scala-fu can figure it out. There is also some cleanup on the imports to remove a Scala -> Java implicit conversion I hadn't realised I was using (and didn't need to use).
